### PR TITLE
A) Implements a complete batch-download of selected files

### PR DESCRIPTION
**Pull request by noVaSon**
Highlights:
• add download Button to SoundInfoModal
• submit freesound-ids to the flask backend
• download the files and the metadata there and put it in a zip archive
• send it back to the frontend and save to disk
• remove the temporary files downloaded from the server with a delayed cleanup Thread

Adresses #38

### CODE REVIEW

**Comment by noVaSon** [comment]
Some notes to myself:

• csv should be optimized (only selected columns, reshape lists (now: [u'tag1, u'tag2']))
• filenames should be identical with the entry in the csv
• parts of the python implementation should be modularized

**Comment by ffont** [comment]
Wow, this looks like big news!
I'll be away for a week approx so I won't be able to review. I'll have a look after that.
In this way you also have time to add further improvements if needed.
Thanks!

**Comment by noVaSon** [comment]
Thank you for this note @ffont!

Seems like we had similar temperatures this summer (37°C in Berlin this week!) :)
Enjoy your vacation!!

**Comment by noVaSon** [comment]
Ok, ready for a merge :D

File: backend/fszipper.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
+from backend.settings import DOWNLOAD_TEMP
```
Comment (ffont) [inline backend/fszipper.py]: `DOWNLOAD_TEMP` need to be added in `settings.example.py`.

Comment (ffont) [reply backend/fszipper.py]: And maybe rename it to `DOWNLOAD_FOLDER_PATH` for consistency with other settings.

File: src/containers/Sidebar/actions.js [hunk 2]
```diff
@@ -9,3 +16,21 @@ export const toggleSidebarVisibility = makeActionCreator(TOGGLE_SIDEBAR_VISIBILI
 export const setSidebarTab = makeActionCreator(SET_SIDEBAR_TAB, 'newTab');
 export const setExampleQueryDone = makeActionCreator(EXAMPLE_QUERY_DONE);
 export const moveSidebarArrow = makeActionCreator(MOVE_SIDEBAR_ARROW, 'position');
+
+export const batchDownloadSelectedOriginals = (selectedSounds, sounds) => (dispatch) => {
+  const fsIds = [];
+  selectedSounds.forEach(soundID => fsIds.push(`${sounds[soundID].id.split('-')[0]}`));
+
+  if (fsIds.length) {
+    dispatch(displaySystemMessage(`Prepare download of ${selectedSounds.length} Sounds, please wait...`, MESSAGE_STATUS.INFO));
```
Comment (ffont) [inline src/containers/Sidebar/actions.js]: "Preparing download of..." better than "Prepare download of..."

**Review by ffont — CHANGES_REQUESTED** [review]
Great work! Here are some comments:

- If not logged in, download fails with 500 error (unauthorized). Maybe this could be prevented by not allowing to download if user not logged in.
- I'd use the same download icon as in the sound panel as it is easier to understand (IMO).
- Also the download icon should disappear when no sounds are selected. 
- I'd add a badge with the number of selected sounds on top of the icon so things become even clearer.
- Building the CSV sometimes fails when trying to download sounds with non ascii characters in their description or name. For example downloading this one fails: https://freesound.org/people/sijeh/sounds/394561/
- Also maybe there is no need to add that much data in the CSV. Could be something simpler like the file we add when a pack is downloaded in Freesound (mainly license/user info)

**Comment by noVaSon** [comment]
News:

I adressed your issues @ffont  - exept the unicode problem creating the .csv
I'll take care of this the next few days.

Best!
Eric

**Review by noVaSon — COMMENTED** [review]
Helpful suggestions! Than you!
It is ready to merge and deploy now.

**Comment by ffont** [comment]
Hi,
I tried it again, it works almost perfect now, but still I have a couple of comments:

- When not logged in, the download still fails. If clicking on download with the user not being logged in we should show a message that this action requires login.
- When selecting multiple sounds (shit + click) I would prevent the info window from opening (for the sound being currently selected). This is not really part of this PR but I realized it while trying it for downloading sounds and to me it makes sense. Basically the idea is that when a sound is selected we show the info window unless shit key is pressed (does this make sense to you?).
- I'd only show the download option when more than 1 sound is selected. Otherwise there is already a download button in the info window of the selected sound. We prefer users to use that method because they'll download directly from Freesound instead of Freesound Explorer downloading from Freesound and then serving to users.
- You'll have to add instructions about the download feature in the README.

Thanks for the work, it's a really nice feature!

**Comment by noVaSon** [comment]
@ffont 

I did it!
Much better now, you are right!

Can you please tell me how to access githubusercontent for uploading pictures though?

**Comment by ffont** [comment]
> Can you please tell me how to access githubusercontent for uploading pictures though?

Hah, I think I played a trick which is to upload an image by dragging it into an issue or PR comment (here!) and then copying the URL. Just trying it now again:

364911_s_a33eb96f_freesoundbeastwhoosh

You drag an image in this editor and Github uploads it and creates a URL than then I used in the readme. Then I guess I deleted the comments. Hacky!

**Comment by noVaSon** [comment]
@ffont > Thanks for the hacky trick 🍡 
I'm DonE!

**Review by ffont — APPROVED** [review]

**Pull request merged**